llvm/lib/Target/AArch64/AArch64InstrInfo.cpp

```diff
@@ -9631,48 +9888,76 @@ AArch64InstrInfo::analyzeLoopForPipelining(MachineBasicBlock *LoopBB) const {
     return nullptr;
 
   // Must be conditional branch
-  if (FBB == nullptr)
+  if (TBB != LoopBB && FBB == nullptr)
     return nullptr;
 
   assert((TBB == LoopBB || FBB == LoopBB) &&
          "The Loop must be a single-basic-block loop");
 
+  MachineInstr *CondBranch = &*LoopBB->getFirstTerminator();
+  const TargetRegisterInfo &TRI = getRegisterInfo();
+
+  if (CondBranch->getOpcode() != AArch64::Bcc)
+    return nullptr;
+
   // Normalization for createTripCountGreaterCondition()
   if (TBB == LoopBB)
     reverseBranchCondition(Cond);
 
-  MachineInstr *CondBranch = &*LoopBB->getFirstTerminator();
-  const TargetRegisterInfo &TRI = getRegisterInfo();
-
-  // Find the immediate predecessor of the conditional branch
-  MachineInstr *PredBranch = nullptr;
-  if (CondBranch->getOpcode() == AArch64::Bcc) {
-    for (MachineInstr &MI : reverse(*LoopBB)) {
-      if (MI.modifiesRegister(AArch64::NZCV, &TRI)) {
-        PredBranch = &MI;
+  MachineInstr *Comp = nullptr;
+  unsigned CompCounterOprNum = 0;
+  for (MachineInstr &MI : reverse(*LoopBB)) {
+    if (MI.modifiesRegister(AArch64::NZCV, &TRI)) {
```

Comment:
Do we need to check for other uses of NZCV?

Reply:
CALL instructions could cause problems because NZCV use/def is not set. It is fixed not to pipeline when CALL or instructions which have unmodeled side effects exist as https://github.com/llvm/llvm-project/pull/65609/commits/0afa899364a4eec2a514984801cb4a740655ee25.

Looking at the code in MachineScheduler, it seems that there are no other instructions that implicitly use NZCV.

https://github.com/llvm/llvm-project/blob/083a26682d1c4d2d6efb051cb2f658564eaa32ff/llvm/lib/CodeGen/ScheduleDAGInstrs.cpp#L550-L555

Reply:
Thanks.